cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```
@@ -0,0 +1,270 @@
+                       "Requested IPC memory handle type not supported");
+
+    ::cudaMemPoolProps __pool_properties{};
+    __pool_properties.allocType     = ::cudaMemAllocationTypePinned;
+    __pool_properties.handleTypes   = ::cudaMemAllocationHandleType(__properties.allocation_handle_type);
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolSetAttribute with cudaMemPoolReuseAllowOpportunistic",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolSetAttribute with cudaMemPoolAttrReleaseThreshold",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__properties.release_threshold);
+
+    // allocate the requested initial size to prime the pool.
+    // We need to use a new stream so we do not wait on other work
+    if (__properties.initial_pool_size != 0)
+    {
+      ::cuda::experimental::stream __temp_stream{__device_id};
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
```

looks like we're using `cudaMallocAsync` even if the driver version is less than `__min_async_version`.

---

Above we have
```cpp
_LIBCUDACXX_ASSERT(_CUDA_VMR::__device_supports_pools(__device_id), "cudaMallocAsync not supported");
```
which should throw an exception if cudaMallocAsync is not supported.